falcon/util/uri.py

```diff
@@ -406,9 +408,31 @@ def unquote_string(quoted):
                           for q in tmp_quoted.split(r'\\')])
 
 
+def decode_wsgi_string(tunneled):
+    """Decode a PEP-3333 WSGI string, such as the request path.
+
+    Args:
+        tunneled (str): The original WSGI string tunneled as Latin-1 bytes as
+            per `PEP-3333 WSGI unicode handling
+            <https://www.python.org/dev/peps/pep-3333/#unicode-issues>`_ notes.
+
+    Returns:
+        str: decoded string
+    """
+    try:
+        # PERF(vytas): Use EAFP here to quickly check if there are any
+        #   non-ASCII bytes in the string, and return the input verbatim if
+        #   possible.
+        tunneled.encode('ascii')
+        return tunneled
+    except ValueError:
+        return tunneled.encode('iso-8859-1').decode('utf-8', 'replace')
```

Comment:
How does performance of this approach compare to [str.isascii()](https://docs.python.org/3/library/stdtypes.html?highlight=isascii#str.isascii)?

Reply:
OMG :man_facepalming: 

I managed to overlook the existence of this method, and reimplemented it. I'll check...

Reply:
Heh, no worries. We'll still need a solution for 3.5 and 3.6 either way.

Reply:
Oh, I got scared it had existed since Stone Age, and I was reimplementing it in Cython :slightly_smiling_face: